backend/tournesol/views/preview.py

```diff
@@ -30,6 +37,65 @@
 COLOR_BROWN_FONT = (29, 26, 20, 255)
 COLOR_NEGATIVE_SCORE = (128, 128, 128, 248)
 
+YT_THUMBNAIL_MQ_SIZE = (320, 180)
+
+
+class PreviewMixin:
+    """
+    A generic mixin that provides common behaviours that can be used by all
+    dynamic preview `APIView`.
+    """
+
+    def default_preview(self):
+        default_preview = open(
+            str(BASE_DIR / "tournesol/resources/tournesol_screenshot_og.png"), "rb"
+        )
+        response = FileResponse(default_preview, content_type="image/png")
+        return response
+
+    def get_entity(self, uid: str) -> Entity:
+        try:
+            entity = Entity.objects.get(uid=uid)
+        except Entity.DoesNotExist as exc:
+            logger.error("Preview impossible entity with UID %s.", uid)
+            logger.error("Exception caught: %s", exc)
+            raise exc
+        return entity
+
+    def is_video(self, entity: Entity) -> None:
+        if entity.type != TYPE_VIDEO:
+            logger.info("Preview not implemented for entity with UID %s.", entity.uid)
+            return False
+        return True
+
+    def get_ts_logo(self, size: tuple):
+        return (
+            Image.open(BASE_DIR / "tournesol/resources/Logo64.png")
+            .convert("RGBA")
+            .resize(size)
+        )
+
+    def get_yt_thumbnail(self, entity: Entity) -> Image:
+        url = f"https://img.youtube.com/vi/{entity.video_id}/mqdefault.jpg"
+        try:
+            thumbnail_response = requests.get(url)
+        except ConnectionError as exc:
```

Comment:
Proposition:

```
class FailedBuildingPreviewError(Exception):
    pass

class AbstractPreviewAPIView(APIView):

    def _get(self, *args, **kwargs):
        raise NotImplementedError()

    def get(self, *args, **kwargs):
        try: 
            self._get(*args, **kwargs)
        except FailedBuildingPreviewError:
            return self.default_preview()
```

Then all the checks do not raise the same error to be caught a second time but instead raise `FailedBuildingPreviewError`. Concerns `get_entity`, `is_video` (could be renamed `ensure_is_video` in case it does raise errors), and `get_yt_thumbnail`

Reply:
You're right, it's indeed unusual.

I've done it on purpose to give the possibility to `get_yt_thumbnail` to catch the error and to log some messages, so that all methods calling `get_yt_thumbnail` don't have to duplicate the log handling.

In a second time, the `DynamicWebsitePreviewEntity.get` method then needs to also catch the error to properly return `self.default_preview()`.

It's a trick, mainly to factorize log messages and to make the real exception available in the view.

If there is no objection, I'd like to keep the code as-is. In our case I prefer raising different types of exception when the root causes are different, to make the code a little bit more explicit.